source/core_mqtt.c

```diff
@@ -2901,7 +3067,12 @@ MQTTStatus_t MQTT_Publish( MQTTContext_t * pContext,
         }
     }
 
-    if( stateUpdateHookExecuted == true )
+    /* mutex should be released and not before updating the state
+     * because we need to make sure that the state is updated
+     * after sending the publish packet, before the receive
+     * loop receives ack for this and would want to update its state
+     */
+    if( stateUpdateHookExecuted == true ) 
```

Comment:
I realize that this was already there, but why not fix it when we are modifying the library for the better :)

Reply:
fixed